File: problem_015.c
// -*- compile-command: "gcc -std=c99 -o problem_015 problem_015.c ceuler.c -Wall -lm" -*-
// Copyright (c) 2014 Michael Caldwell
#include <stdio.h>
#include "ceuler.h"

int main()
{
    printf("Project Euler - Problem 15:\n"
           "Starting in the top left corner in a 20 by 20 grid, how many routes are there to the bottom right corner?\n\n");

    // Begin time tracking
    struct timeval start;
    gettimeofday(&start, NULL);

    unsigned long long pascal[100][100] = {};
    int n = 0, i = 0;
    bool debug = false;

    pascal[0][0]=1;

    while (i < 40)
    {
        i++;
        for (n=0 ; n<100 ; n++)
            if (n-1 < 0)
                pascal[i][n]=pascal[i-1][n];
            else
                pascal[i][n]=pascal[i-1][n]+pascal[i-1][n-1];

        // DEBUG - Print row
        if (debug == true)
        {
            printf("%2i: ", i);
            for (n=0 ; n<100 ; n++)
                if (pascal[i][n]==0)
                    break;
                else
                    printf("%6llu ", pascal[i][n]);
            printf("\n\n");
        }
    }

    printf("Answer: %llu\n", pascal[40][20]);
    printElapsedTime(start);

    return 0;
}

File: problem_015.go
// -*- compile-command: "go build -o problem_015_go problem_015.go" -*-
// Copyright (c) 2017 Michael Caldwell
package main

import "fmt"

func main() {
	fmt.Println("Project Euler - Problem 15")
	fmt.Println("Starting in the top left corner in a 20 by 20 grid, how many routes are there to the bottom right corner?\n")

	pascal := [][]int{{1, 0}}
	limit := 20
	debug := false

	// Create Pascal's Triangle
	for y := 0; y < limit * 2; y++ {
		pascal = append(pascal, make([]int, len(pascal[y]) + 1))
		pascal[y+1][0] = 1
		for x := 1; x < len(pascal[y]); x++ {
			pascal[y+1][x] = pascal[y][x-1] + pascal[y][x]
		}
	}

	// Print entire triangle
	if debug == true {
		for y := 0; y < limit * 2; y++ {
			for x := 0; x < len(pascal[y])-1; x++ {
				fmt.Printf("%d ", pascal[y][x])
			}
			fmt.Println("")
		}
	}

	fmt.Println(pascal[limit*2][limit])
}
